Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp

```diff
@@ -0,0 +1,185 @@
+		if (SingletonEntityId == SpatialConstants::INVALID_ENTITY_ID)
+		{
+			continue;
+		}
+
+		AActor* SingletonActor = nullptr;
+		USpatialActorChannel* Channel = nullptr;
+		GetSingletonActorAndChannel(Pair.Key, SingletonActor, Channel);
+
+		// Singleton wasn't found or channel is already set up
+		if (Channel == nullptr || Channel->Actor != nullptr)
+		{
+			continue;
+		}
+
+		// Add to entity registry
+		// This indirectly causes SetChannelActor to not create a new entity for this actor
+		NetDriver->GetEntityRegistry()->AddToRegistry(SingletonEntityId, SingletonActor);
+
+		Channel->SetChannelActor(SingletonActor);
+
+
+		UnrealMetadata* Metadata = View->GetUnrealMetadata(SingletonEntityId);
+		if (Metadata == nullptr)
+		{
+			// Don't have entity checked out
+			continue;
+		}
+
+		// Since the entity already exists, we have to handle setting up the PackageMap properly for this Actor
+		NetDriver->PackageMap->ResolveEntityActor(SingletonActor, SingletonEntityId, Metadata->SubobjectNameToOffset);
+		UE_LOG(LogTemp, Log, TEXT("Linked Singleton Actor %s with id %d"), *SingletonActor->GetClass()->GetName(), SingletonEntityId);
+	}
+}
+
+void UGlobalStateManager::ExecuteInitialSingletonActorReplication()
+{
+	for (const auto& pair : SingletonNameToEntityId)
+	{
+		Worker_EntityId SingletonEntityId = pair.Value;
+
+		// Entity has already been created on another server
+		if (SingletonEntityId != SpatialConstants::INVALID_ENTITY_ID)
+		{
+			continue;
+		}
+
+		AActor* SingletonActor = nullptr;
+		USpatialActorChannel* Channel = nullptr;
+		GetSingletonActorAndChannel(pair.Key, SingletonActor, Channel);
+
+		// Class couldn't be found
+		if (Channel == nullptr)
+		{
+			continue;
+		}
+
+		// Set entity id of channel from the GlobalStateManager.
+		// If the id was 0, SetChannelActor will create the entity.
+		// If the id is not 0, it will start replicating to that entity.
+		Channel->SetChannelActor(SingletonActor);
+
+		UE_LOG(LogTemp, Log, TEXT("Started replication of Singleton Actor %s"), *SingletonActor->GetClass()->GetName());
+	}
+}
+
+void UGlobalStateManager::UpdateSingletonEntityId(const FString& ClassName, const Worker_EntityId SingletonEntityId)
+{
+	SingletonNameToEntityId[ClassName] = SingletonEntityId;
+
+	Worker_ComponentUpdate Update = {};
+	Update.component_id = SpatialConstants::GLOBAL_STATE_MANAGER_COMPONENT_ID;
+	Update.schema_type = Schema_CreateComponentUpdate(SpatialConstants::GLOBAL_STATE_MANAGER_COMPONENT_ID);
+	Schema_Object* UpdateObject = Schema_GetComponentUpdateFields(Update.schema_type);
+
+	Schema_AddStringToEntityMap(UpdateObject, 1, SingletonNameToEntityId);
+
+	NetDriver->Connection->SendComponentUpdate(SpatialConstants::GLOBAL_STATE_MANAGER, &Update);
+}
+
+void UGlobalStateManager::GetSingletonActorAndChannel(FString ClassName, AActor*& OutActor, USpatialActorChannel*& OutChannel)
+{
+	OutActor = nullptr;
+	OutChannel = nullptr;
+
+	UClass* SingletonActorClass = FindObject<UClass>(ANY_PACKAGE, *ClassName);
```

Comment:
This could potentially fail for blueprint singleton objects. We should be doing a LoadObject instead.